### PR TITLE
Survival Extension

### PR DESCRIPTION
**Pull request by carobec**
add an extension to enable `mcboost` to deal with survival models

### CODE REVIEW

File: R/MCBoost.R [hunk 1]
```diff
@@ -386,19 +360,73 @@ MCBoost = R6::R6Class("MCBoost",
...
+        }))
+        buckets[[2]]$lower = -Inf
+        buckets[[length(buckets)]]$upper = Inf
+      } else {
+        if (self$num_buckets == 1L) stop("If partition=TRUE, num_buckets musst be > 1!")
+      }
+      buckets
+    }, 
+    
+    assert_prob = function(prob, data) {
```
Comment (pfistfl) [inline R/MCBoost.R:420]: could be a normal helper function, not  part of the class.
Then it might also check 0 <= x <= 1

File: R/MCBoostSurv.R [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+      multiplicative=TRUE,
+      auditor_fitter=NULL,
+      subpops=NULL,
+      default_model_class=ConstantPredictor, #FIXME must be constant over time
+      init_predictor=NULL,
+      iter_sampling="none") {
+      
+      #FIXME does it make sense to include times here already? Preprocess? Checks? 
+      
+      self$times = times
```
Comment (pfistfl) [inline R/MCBoostSurv.R:27]: perhaps `time_points`?

File: R/MCBoost.R (hunk 1 above)
Comment (carobec) [reply R/MCBoost.R:423]: @pfistfl 
This function is only used in public functions of `MCBoost`. Depending on the class (`MCBoost` or `MCBoostSurv`)  `prob` has to be a different type and `assert_prob` has to do something else. (In classification a vector of probabilities - numeric values between 0 & 1 - and in survival either a `distr6 `object or a matrix with different probabilities.) That was my idea.

How would you put it to the helper functions? Have the class of the R6 object as an input? :)

File: R/MCBoostSurv.R [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+
+    # calculate for every time step and every survival curve the residual
+    calc_residual_matrix_r = function(prediction, labels) {
+      nr_obs = length(labels)
+      nc_times = length(self$time_points)
+      labels_num = as.numeric(labels)
+
+      igs = matrix(nrow = nr_obs, ncol = nc_times)
+
+      for (i in seq_len(nr_obs)) {
```
Comment (pfistfl) [inline R/MCBoostSurv.R:148]: Rewrite as 
```r
igs = prediction
mask = outer(labels_num, self$time_points, FUN = ">")
igs[mask] = igs[mask]  - 1
```

Comment (pfistfl) [reply R/MCBoostSurv.R:148]: Should be drastically faster

File: R/helpers.R [hunk 4]
```diff
@@ -62,7 +66,30 @@ mlr3_init_predictor = function(learner) {
...
+  }
+}
+
+
+even_bucket = function(pos, frac, min, max) {
+  min + pos / frac * (max - min)
+}
+
+
+make_survival_curve = function (prediction){
```
Comment (pfistfl) [inline R/helpers.R:80]: This looks pretty inefficient.
1. Can we rewrite as `apply(prediction, 1, function(x) {cm = cummin(x); if(x != cm) message(...); x = cm else x})`
or so?
2.Let us think whether we can do some smoothing at some point in the future.

File: R/MCBoostSurv.R (hunk 3 above)
Comment (carobec) [reply R/MCBoostSurv.R]: True, thank you!

File: R/helpers.R (hunk 4 above)
Comment (carobec) [reply R/helpers.R]: Thank you, also applied this is in the latest commit. 
Smoothing is a good point.. :) I'm writing it down.

**Pull request merged**